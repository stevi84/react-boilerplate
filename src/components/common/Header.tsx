import { LanguageSwitch } from './LanguageSwitch';
import {
  AppBar,
  Divider,
  IconButton,
  ListItemIcon,
  ListItemText,
  Menu,
  MenuItem,
  Toolbar,
  Typography,
} from '@mui/material';
import MenuIcon from '@mui/icons-material/Menu';
import { useState } from 'react';
import Home from '@mui/icons-material/Home';
import People from '@mui/icons-material/People';
import EditNote from '@mui/icons-material/EditNote';
import { useNavigate } from 'react-router-dom';
import { getEnv } from '../../globals/Environments';

export const Header = () => {
  const navigate = useNavigate();
  const [anchorEl, setAnchorEl] = useState<HTMLElement | null>(null);
  const open = Boolean(anchorEl);
  const handleClick = (event: React.MouseEvent<HTMLButtonElement>) => {
    setAnchorEl(event.currentTarget);
  };
  const handleClose = () => {
    setAnchorEl(null);
  };

  return (
    <>
      <AppBar position="fixed">
        <Toolbar>
          <IconButton size="large" edge="start" color="inherit" onClick={handleClick} sx={{ mr: 2 }}>
            <MenuIcon />
          </IconButton>
          <Menu id="menu" anchorEl={anchorEl} open={open} onClose={handleClose}>
            <MenuItem onClick={() => navigate('/')}>
              <ListItemIcon>
                <Home fontSize="small" />
              </ListItemIcon>
              <ListItemText>Home</ListItemText>
            </MenuItem>
            <Divider />
            <MenuItem onClick={() => navigate('/todos')}>
              <ListItemIcon>
                <EditNote fontSize="small" />
              </ListItemIcon>
              <ListItemText>Todos</ListItemText>
            </MenuItem>
            <MenuItem onClick={() => navigate('/users')}>
              <ListItemIcon>
                <People fontSize="small" />
              </ListItemIcon>
              <ListItemText>Users</ListItemText>
            </MenuItem>
          </Menu>
          <Typography variant="h6" component="div" sx={{ flexGrow: 1 }}>
            React Boilerplate
          </Typography>
          <Typography variant="body1" component="div" sx={{ mr: 1 }}>
            Environment: {getEnv()}
          </Typography>
          <LanguageSwitch />
        </Toolbar>
      </AppBar>
      <Toolbar />
    </>
  );
};
