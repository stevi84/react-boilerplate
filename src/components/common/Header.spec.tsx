import { describe, it, expect, vi } from 'vitest';
import { fireEvent, render, screen } from '@testing-library/react';
import { Header } from './Header';

const { navigateMock } = vi.hoisted(() => ({ navigateMock: vi.fn() }));
vi.mock('react-router-dom', () => ({ useNavigate: () => navigateMock }));

vi.mock('../../globals/Environments', () => ({ getEnv: () => 'testenv' }));

describe('Header', () => {
  it('should equal saved snapshot', () => {
    const tree = render(<Header />).asFragment();
    expect(tree).toMatchSnapshot();
  });

  it('should show the environment', () => {
    render(<Header />);
    expect(screen.getByText(/testenv/)).toBeTruthy();
  });

  it('should navigate to dashboard', () => {
    render(<Header />);
    expect(navigateMock.mock.calls.length).toEqual(0);
    fireEvent.click(screen.getByTestId('MenuIcon'));
    fireEvent.click(screen.getByText('Home'));
    expect(navigateMock.mock.calls.length).toEqual(1);
    expect(navigateMock.mock.calls[0][0]).toEqual('/');
  });

  it('should navigate to users', () => {
    render(<Header />);
    expect(navigateMock.mock.calls.length).toEqual(0);
    fireEvent.click(screen.getByTestId('MenuIcon'));
    fireEvent.click(screen.getByText('Users'));
    expect(navigateMock.mock.calls.length).toEqual(1);
    expect(navigateMock.mock.calls[0][0]).toEqual('/users');
  });

  it('should navigate to todos', () => {
    render(<Header />);
    expect(navigateMock.mock.calls.length).toEqual(0);
    fireEvent.click(screen.getByTestId('MenuIcon'));
    fireEvent.click(screen.getByText('Todos'));
    expect(navigateMock.mock.calls.length).toEqual(1);
    expect(navigateMock.mock.calls[0][0]).toEqual('/todos');
  });
});
