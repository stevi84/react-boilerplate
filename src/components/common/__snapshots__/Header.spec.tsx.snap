// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Header > should equal saved snapshot 1`] = `
<DocumentFragment>
  <header
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed css-11b3ww9-MuiPaper-root-MuiAppBar-root"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeLarge css-ancrnh-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="MenuIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <div
        class="MuiTypography-root MuiTypography-h6 css-1juivf6-MuiTypography-root"
      >
        React Boilerplate
      </div>
      <div
        class="MuiTypography-root MuiTypography-body1 css-1fbouff-MuiTypography-root"
      >
        Environment: testenv
      </div>
      <div
        class="MuiBox-root css-ece9u5"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
          >
            language
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
          >
            <div
              aria-controls=":r0:"
              aria-expanded="false"
              aria-haspopup="listbox"
              class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
              role="combobox"
              tabindex="0"
            >
              <span
                class="notranslate"
              >
                â€‹
              </span>
            </div>
            <input
              aria-hidden="true"
              aria-invalid="false"
              class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
              tabindex="-1"
              value=""
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  language
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
  </header>
  <div
    class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
  />
</DocumentFragment>
`;
